src/postoffice.cc

```diff
@@ -25,78 +25,104 @@ Postoffice::Postoffice() {
+                    kWorkerGroup + kScheduler,
+                    kWorkerGroup + kServerGroup + kScheduler}) {
+        node_ids_[g].push_back(id);
+      }
     }
-  }
 
-  for (int i = 0; i < num_servers_; ++i) {
-    int id = ServerRankToID(i);
-    for (int g : {id, kServerGroup, kWorkerGroup + kServerGroup,
-            kServerGroup + kScheduler,
-            kWorkerGroup + kServerGroup + kScheduler}) {
-      node_ids_[g].push_back(id);
+    for (int i = 0; i < num_servers_; ++i) {
+      int id = ServerRankToID(i);
+      for (int g : {id, kServerGroup, kWorkerGroup + kServerGroup,
+                    kServerGroup + kScheduler,
+                    kWorkerGroup + kServerGroup + kScheduler}) {
+        node_ids_[g].push_back(id);
+      }
     }
-  }
 
-  for (int g : {kScheduler, kScheduler + kServerGroup + kWorkerGroup,
-          kScheduler + kWorkerGroup, kScheduler + kServerGroup}) {
-    node_ids_[g].push_back(kScheduler);
+    for (int g : {kScheduler, kScheduler + kServerGroup + kWorkerGroup,
+                  kScheduler + kWorkerGroup, kScheduler + kServerGroup}) {
+      node_ids_[g].push_back(kScheduler);
+    }
+    init_stage++;
   }
+  start_mu_.unlock();
 
   // start van
-  van_->Start();
-
-  // record start time
-  start_time_ = time(NULL);
-
-  // do a barrier here
-  if (do_barrier) Barrier(kWorkerGroup + kServerGroup + kScheduler);
+  van_->Start(customer_id);
+
+  start_mu_.lock();
+  if (init_stage == 1) {
+    // record start time
+    start_time_ = time(NULL);
+    // do a barrier here
+    init_stage++;
+  }
+  start_mu_.unlock();
+  if (do_barrier) Barrier(customer_id, kWorkerGroup + kServerGroup + kScheduler);
 }
 
-void Postoffice::Finalize(const bool do_barrier) {
-  if (do_barrier) Barrier(kWorkerGroup + kServerGroup + kScheduler);
-  van_->Stop();
-  if (exit_callback_) exit_callback_();
+void Postoffice::Finalize(const int customer_id, const bool do_barrier) {
+  if (do_barrier) Barrier(customer_id, kWorkerGroup + kServerGroup + kScheduler);
+  if (customer_id == 0) {
+    van_->Stop();
+    if (exit_callback_) exit_callback_();
+  }
 }
 
 
 void Postoffice::AddCustomer(Customer* customer) {
   std::lock_guard<std::mutex> lk(mu_);
-  int id = CHECK_NOTNULL(customer)->id();
-  CHECK_EQ(customers_.count(id), (size_t)0) << "id " << id << " already exists";
-  customers_[id] = customer;
+  int app_id = CHECK_NOTNULL(customer)->app_id();
+  if (customers_.find(app_id) == customers_.end()) {
+    customers_[app_id] = *(new std::unordered_map<int, Customer*>());
+  }
+  // check if the customer id has existed
+  int customer_id = CHECK_NOTNULL(customer)->customer_id();
+  CHECK_EQ(customers_[app_id].count(customer_id), (size_t)0) << "customer_id " \
+    << customer_id << " already exists\n";
+  customers_[app_id][customer_id] = customer;
+  if (barrier_done_.find(app_id) == barrier_done_.end()) {
+    barrier_done_[app_id] = *(new std::unordered_map<int, bool>());
```

Comment:
memory leak

Reply:
how? please forgive that I have nearly forgotten everything of c++, but this should be alive as long as the whole process